astropy/table/sorted_array.py

```diff
@@ -0,0 +1,286 @@
+import numpy as np
+
+class SortedArray(object):
+    '''
+    Implements a sorted array container using
+    a list of numpy arrays.
+
+    Parameters
+    ----------
+    lines : `Table`
+        Sorted columns of the original table as well as
+        a final column consisting of the rows given by argsort()
+    col_dtypes : list or None
+        If not None, col_dtypes should be a list of types for each column
+    data : list or None
+        An internal data object to use instead of creating new data
+    '''
+    def __init__(self, lines, col_dtypes=None, data=None):
+        if data is not None: # sliced reference to data
+            self.length = len(data[0])
+            self._data = data
+            return
+        elif col_dtypes is not None:
+            self.num_cols = len(col_dtypes)
+        else:
+            self.num_cols = len(lines.colnames) - 1
+            col_dtypes = [x.info.dtype for x in lines.columns[:-1].values()]
+
+        self.length = 0 if lines is None else len(lines)
+        self._data = [np.zeros(self.length, dtype=d) for d in col_dtypes]
```

Comment:
And it's not completely clear that you even need to copy since the `data` that get passed in are a sorted version of the original columns so it is already a clean copy that isn't going to be manipulated elsewhere.  In general I wouldn't guess that people should be creating `Index` objects themselves so we can afford to be less conservative about copying.